### PR TITLE
[Merged by Bors] - feat: port Testing.SlimCheck.Functions

### PR DESCRIPTION
**Pull request by kim-em**
This diverges fairly considerably from mathlib3 since we changed the `SampleableExt` API when porting the previous file.

---

- [x] depends on: #5796

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Control/ULiftable.lean [hunk 1]
```diff
@@ -110,6 +110,9 @@ open ULift
 instance : ULiftable id id where
   congr F := F
 
+instance : ULiftable Id Id where
+  congr F := F
+
```
Comment (eric-wieser) [inline Mathlib/Control/ULiftable.lean:115]: Why do we have two different spellings available here?

File: Mathlib/Testing/SlimCheck/Gen.lean [hunk 2]
```diff
@@ -73,10 +74,12 @@ def getSize : Gen Nat :=
 def resize (f : Nat → Nat) (x : Gen α) : Gen α :=
   withReader (ULift.up ∘ f ∘ ULift.down) x
 
+variable {α : Type u}
```
Comment (eric-wieser) [inline Mathlib/Testing/SlimCheck/Gen.lean:77]: Auto-implicits were making this error invisible.

File: Mathlib/Testing/SlimCheck/Functions.lean [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+  proxy := TotalFunction α (SampleableExt.proxy β)
+  interp f := SampleableExt.interp ∘ f.apply
+  sample := do
+    let xs : List (_ × _) ← (SampleableExt.sample (α := List (α × β)))
+    let ⟨x⟩ ← (ULiftable.up <|
+      SampleableExt.sample : Gen (ULift.{max u ub} (SampleableExt.proxy β)))
+    pure <| TotalFunction.withDefault (List.toFinmap' <| xs.map <|
+      Prod.map SampleableExt.interp id) x
+  -- note: no way of shrinking the domain without an inverse to `interp`
+  shrink := { shrink := letI : Shrinkable α := {}; TotalFunction.shrink }
```
Comment (eric-wieser) [inline Mathlib/Testing/SlimCheck/Functions.lean:157]: I'm not sure what the right thing to do is here: we only have `Shrinkable (proxy α)` available, but there's no easy way to work with `TotalFunction (SampleableExt.proxy α) (SampleableExt.proxy β)` instead

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5796
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5796~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Review by jcommelin — APPROVED** [review]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

**Comment by eric-wieser** [comment]
CI won't pass, this has some minor lint errors that I won't have time to get to today.

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**